test-network-function/networking/suite.go

```diff
@@ -275,3 +300,123 @@ func testNodePort(env *config.TestEnvironment) {
+		noderes := PortDeclared{port: p, name: name, protocol: protocolName}
+		result = append(result, noderes)
+		port, name, protocolName = "", "", ""
+	}
+	return result
+}
+
+func getListeningPortList(temp []string) []PortListen {
+	listeningPortCommand := strings.Join(temp, " ")
+	fmt.Println(listeningPortCommand)
+	res, _ := utils.ExecuteCommand(listeningPortCommand, ocCommandTimeOut, interactive.GetContext(false))
+	lines := strings.Split(res, "\n")
+	var result []PortListen
+	for _, line := range lines {
+		fields := strings.Fields(line)
+		if indexProtocolName > len(fields) || indexPort > len(fields) {
+			return result
+		}
+		s := strings.Split(fields[indexPort], ":")
+		p, _ := strconv.Atoi(s[1])
+		nodeListen := PortListen{port: p, protocol: fields[indexProtocolName]}
+		result = append(result, nodeListen)
+	}
+	return result
+}
+
+func checkIfListenIsDeclared(declaredPortList []PortDeclared, listeningPortList []PortListen) bool {
+	indexElement1 := 0
+	indexElement2 := 0
+
+	for indexElement1 < len(declaredPortList) && indexElement2 < len(listeningPortList) {
+		if declaredPortList[indexElement1].port == listeningPortList[indexElement2].port {
+			indexElement1++
+			indexElement2++
+			continue
+		}
+		if listeningPortList[indexElement2].port < declaredPortList[indexElement1].port {
+			return false
+		}
+		indexElement1++
+	}
+	return true
+}
+
+func testListenAndDeclared(env *config.TestEnvironment) {
+	var declaredPort []PortDeclared
+	var listeningPort []PortListen
+	var x *config.Container
+	res := true
+	testID := identifiers.XformToGinkgoItIdentifier(identifiers.TestServicesDoNotUseNodeportsIdentifier)
+	ginkgo.It(testID, func() {
+		for _, podUnderTest := range env.PodsUnderTest {
+			for i := 0; i < podUnderTest.ContainerCount; i++ {
+				temp := getDeclaredPortList(i, podUnderTest.Name, podUnderTest.Namespace)
+				declaredPort = append(declaredPort, temp...)
+			}
+
+			nodeName := podnodename.NewPodNodeName(common.DefaultTimeout, podUnderTest.Name, podUnderTest.Namespace)
+			context := common.GetContext()
+			test, err := tnf.NewTest(context.GetExpecter(), nodeName, []reel.Handler{nodeName}, context.GetErrorChannel())
+			gomega.Expect(err).To(gomega.BeNil())
+			test.RunAndValidate()
+			nodeOc := env.NodesUnderTest[nodeName.GetNodeName()].DebugContainer.GetOc()
+			for _, cut := range env.ContainersUnderTest {
+				fmt.Println(cut)
+				if cut.ContainerIdentifier.PodName == podUnderTest.Name && cut.ContainerIdentifier.Namespace == podUnderTest.Namespace {
+					x = cut
+					break
+				}
+			}
+			containerPID := utils.GetContainerPID(nodeName.GetNodeName(), nodeOc, x.ContainerUID, x.ContainerRuntime)
+
+			temp := []string{utils.AddNsenterPrefix(containerPID), CommandPortListen, "-tulwnH"}
+			fmt.Println(temp)
+			temp1 := getListeningPortList(temp)
+			listeningPort = append(listeningPort, temp1...)
+
+			// sorting the lists
+			sort.SliceStable(declaredPort, func(i, j int) bool {
```

Comment:
compare and sort should include protocol too

Reply:
better move the sorting into the compare function and make some good unit test for it, i.e. different port/protocol combinations in different orders

Reply:
I would add on this, it's lot easier to compare maps than to compare ports. 
in all cases, a good unit test as Jun suggested is required.

Reply:
I agree and I fixed it.